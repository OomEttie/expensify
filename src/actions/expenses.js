const uuidV4 = require('uuid/v4');
import database from '../firebase/firebase';

//get the correct expenses db ref path
let expensesRef = uid => {
  return 'users/' + uid + '/expenses';
};

//
// ADD_EXPENSES
//
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = { description, note, amount, createdAt };

    const dbPath = expensesRef(
      getState().auth ? getState().auth.uid : 'UID_unknown'
    );
    
    return database
      .ref(dbPath)
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      });
  };
};

//
// REMOVE_EXPENSE
//
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id: id
});

export const startRemoveExpense = ({ id }) => {
  return (dispatch, getState) => {
    const dbPath = expensesRef(
      getState().auth ? getState().auth.uid : 'UID_unknown'
    );
    return database
      .ref(`${dbPath}/${id}`)
      .remove()
      .then(data => {
        dispatch(removeExpense({ id }));
      });
  };
};

//
// EDIT_EXPENSE
//
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const dbPath = expensesRef(
      getState().auth ? getState().auth.uid : 'UID_unknown'
    );
    return database
      .ref(`${dbPath}/${id}`)
      .update({
        ...updates
      })
      .then(() => {
        dispatch(editExpense(id, updates));
      });
  };
};

//
// SET_EXPENSES
export const setExpenses = expenses => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const dbPath = expensesRef(
      getState().auth ? getState().auth.uid : 'UID_unknown'
    );
    return database
      .ref(dbPath)
      .once('value')
      .then(snapshot => {
        const expenses = [];

        snapshot.forEach(snapshotChild => {
          expenses.push({
            id: snapshotChild.key,
            ...snapshotChild.val()
          });
        });
        dispatch(setExpenses(expenses));
      })
      .catch(e => {});
  };
};
