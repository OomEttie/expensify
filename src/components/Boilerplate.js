import React from 'react';
import Header from './Header';

export default class BoilerplateApp extends React.Component {
  state = {
  };

  //FUNCTIONS
  render() {
    return (
      <div>
        <Header />
      </div>
    );
  }
}
