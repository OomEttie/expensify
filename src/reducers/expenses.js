const expensesReducerDefaultState = [];

export const expensesReducer = (
  state = expensesReducerDefaultState,
  action
) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense];
    // return state.concat(action.expense);
    case 'REMOVE_EXPENSE':
      return state.filter(data => data.id != action.id);
    case 'EDIT_EXPENSE': {
      // const expenseIndex = state.findIndex(data => data.id == action.id);
      // return { ...state[expenseIndex], amount: action.amount };
      return state.map(expense => {
        // console.log('expense-', expense);
        // console.log('EDIT_EXPENSE action-', action.id);
        if (expense.id == action.id) {
          return { ...expense, ...action.updates };
        } else {
          return expense;
        }
      });
    }
    case 'SET_EXPENSES':
      return action.expenses;
    default:
      return state;
  }
};
