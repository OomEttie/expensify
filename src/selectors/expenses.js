import moment from 'moment';

export const getVisibleExpenses = (
  expenses,
  { text, sortBy, startDate, endDate }
) => {
  return expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, 'day')
        : true;
      // typeof startDate != 'number' || expense.createdAt >= startDate;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, 'day')
        : true;
      // typeof endDate != 'number' || expense.createdAt <= endDate;
      const textMatch =
        expense.description.toLowerCase().indexOf(text.toLowerCase()) >= 0;

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy == 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      }
      if (sortBy == 'amount') {
        return a.amount > b.amount ? -1 : 1;
      }
    });
};
