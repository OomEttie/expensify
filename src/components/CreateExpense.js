import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';
// onSubmit = expense => {
//   console.log('onSubmit-', expense);
// };

export class CreateExpense extends React.Component {
  onSubmit = expense => {
    // props.dispatch(addExpense(expense));
    this.props.createExpense(expense);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <h3>Add Expense</h3>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => {
  return {
    createExpense: expense => {
      dispatch(addExpense(expense));
    }
  };
};

export default connect(
  undefined,
  mapDispatchToProps
)(CreateExpense);
