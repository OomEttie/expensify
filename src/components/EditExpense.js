import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpense extends React.Component {
  onSubmit = expense => {
    // props.dispatch(addExpense(expense));
    this.props.startEditExpense(this.props.expense.id, { ...expense });
    this.props.history.push('/');
  };

  onClick = () => {
    // console.log(dispatch);
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
          <button className="button__warning" onClick={this.onClick}>REMOVE</button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  // console.log('mapStateToProps-props->', props)
  return {
    expense: state.expenses.find(expense => {
      return expense.id == props.match.params.id;
    })
  };
};

const mapDispatchToProps = dispatch => {
  return {
    startEditExpense: (id, updates) => {
      dispatch(startEditExpense(id, updates));
    },
    startRemoveExpense: id => {
      dispatch(startRemoveExpense(id));
    }
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpense);
