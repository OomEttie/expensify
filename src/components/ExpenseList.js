import React from 'react';
import { connect } from 'react-redux';
import { ExpenseListItem } from './ExpenseListItem';
import { getVisibleExpenses } from '../selectors/expenses';

// <h1>Expense List</h1>
export const ExpenseList = props => (
  <div>
    {props.expenses.length == 0 ? (
      <p>No Expenses</p>
    ) : (
      props.expenses.map(expense => {
        // console.log(expense);
        // const Temp = () => (
        //   <div>
        //     <p>Amount: </p>
        //     <p>Created At: </p>
        //   </div>
        // )
        // return <Temp />;
        return <ExpenseListItem key={expense.id} {...expense} />;
      })
    )}
  </div>
);

const mapStateToProps = state => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters)
  };
};

const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);

// export default ExpenseList;
export default ConnectedExpenseList;
