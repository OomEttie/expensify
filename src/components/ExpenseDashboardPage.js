import React from 'react';
import { connect } from 'react-redux';

import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import { getVisibleExpenses } from '../selectors/expenses';
import { getTotalExpensesAmount } from '../selectors/expenses-total';

export const ExpenseDashboardPage = props => (
  <div>
    <h3>This is my expense Dashboard component!</h3>
    <h3>
      Viewing {props.expenses.length} expenses totalling
      {getTotalExpensesAmount(props.expenses)}
    </h3>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

const mapStateToProps = state => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters)
  };
};

export const ConnectedExpenseDashboardPage = connect(mapStateToProps)(
  ExpenseDashboardPage
);

// export default ExpenseDashboardPage;
// export default ConnectedExpenseDashboardPage;
