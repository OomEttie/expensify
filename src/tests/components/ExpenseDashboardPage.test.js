import React from 'react';
import { shallow } from 'enzyme';

import { ExpenseDashboardPage } from '../../components/ExpenseDashboardPage';
import expenses from '../fixtures/expenses';

test('render dashboard component', () => {
  // const store = { expenses: expenses };
  // const store = {
  //   subscribe: () => {},
  //   dispatch: () => {},
  //   getState: () => ({
  //     expenses: expenses,
  //     filters: {
  //       text: '',
  //       sortBy: 'date',
  //       startDate: undefined,
  //       endDate: undefined
  //     }
  //   })
  // };

  const wrapper = shallow(<ExpenseDashboardPage expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});
