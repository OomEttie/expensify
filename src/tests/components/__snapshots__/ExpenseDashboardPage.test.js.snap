// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render dashboard component 1`] = `
<div>
  <h3>
    This is my expense Dashboard component!
  </h3>
  <h3>
    Viewing 
    3
     expenses totalling
    114195
  </h3>
  <Connect(ExpenseListFilters) />
  <Connect(ExpenseList) />
</div>
`;
